### PR TITLE
define the interface for the thread pool

### PR DESCRIPTION
**Pull request by inabao**
#253

### CODE REVIEW

File: include/vsag/thread_pool.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      *              No additional threads will be created beyond this limit.
+      */
+    virtual void
+    set_pool_size(std::size_t limit) = 0;
+
+    /**
+      * Destructor.
+      *
+      * Cleans up resources used by the thread pool.
+      */
```
Comment (wxyucs) [inline include/vsag/thread_pool.h]: public method's name should start with a capital letter

File: include/vsag/thread_pool.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      * @return std::future<void> A future object representing the
+      *                           asynchronous execution of the task,
+      *                           which can be used to obtain the task's
+      *                           result and status.
+      */
+    virtual std::future<void>
+    enqueue(std::function<void(void)> task) = 0;
+};
+
+}  // namespace vsag
```
Comment (wxyucs) [inline include/vsag/thread_pool.h]: add a newline at EOF

File: src/default_thread_pool.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+DefaultThreadPool::set_pool_size(std::size_t limit) {
+    pool_->set_pool_size(limit);
+}
+
+void
+DefaultThreadPool::set_queue_size_limit(std::size_t limit) {
+    pool_->set_queue_size_limit(limit);
+}
+
+}  // namespace vsag
```
Comment (wxyucs) [inline src/default_thread_pool.cpp]: add a newline at EOF

File: src/default_thread_pool.cpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+void
+DefaultThreadPool::wait_until_empty() {
+    pool_->wait_until_empty();
+}
+
+void
+DefaultThreadPool::set_pool_size(std::size_t limit) {
+    pool_->set_pool_size(limit);
```
Comment (wxyucs) [inline src/default_thread_pool.cpp:36]: Is it safe to update the pool size while some tasks are executing

File: src/default_thread_pool.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    set_queue_size_limit(std::size_t limit) override;
+
+    void
+    set_pool_size(std::size_t limit) override;
+
+private:
+    std::unique_ptr<progschj::ThreadPool> pool_;
+};
+
+}  // namespace vsag
```
Comment (wxyucs) [inline src/default_thread_pool.h]: add a newline at EOF

File: src/default_thread_pool.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    enqueue(std::function<void(void)> task) override;
+
+    void
+    wait_until_empty() override;
+
+    void
+    set_queue_size_limit(std::size_t limit) override;
+
+    void
+    set_pool_size(std::size_t limit) override;
```
Comment (wxyucs) [inline src/default_thread_pool.h]: public method's name should start with a capital letter

File: src/safe_thread_pool.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+// limitations under the License.
+
+#pragma once
+
+#include "default_thread_pool.h"
+#include "logger.h"
+
+namespace vsag {
+
+class SafeThreadPool : public ThreadPool {
```
Comment (wxyucs) [inline src/safe_thread_pool.h:23]: why we need a safe thread pool?

**Review by wxyucs — CHANGES_REQUESTED** [review]

File: include/vsag/thread_pool.h (hunk 1 above)
Comment (inabao) [reply include/vsag/thread_pool.h]: done

File: include/vsag/thread_pool.h (hunk 2 above)
Comment (inabao) [reply include/vsag/thread_pool.h]: done

File: src/default_thread_pool.cpp (hunk 3 above)
Comment (inabao) [reply src/default_thread_pool.cpp]: done

File: src/default_thread_pool.h (hunk 5 above)
Comment (inabao) [reply src/default_thread_pool.h]: done

File: src/safe_thread_pool.h (hunk 7 above)
Comment (inabao) [reply src/safe_thread_pool.h:23]: Similar to the allocator, use safe_thread_pool to manage the externally provided thread_pool and the internally created thread_pool.

File: src/default_thread_pool.h (hunk 6 above)
Comment (inabao) [reply src/default_thread_pool.h]: done

File: src/default_thread_pool.cpp [hunk 8]
```diff
@@ -0,0 +1,33 @@
+//
+// Created by root on 12/22/24.
+//
+
```
Comment (wxyucs) [inline src/default_thread_pool.cpp]: replace with license

File: src/safe_thread_pool.h [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    SetPoolSize(std::size_t limit) override {
+        pool_->SetPoolSize(limit);
+    }
+
+private:
+    ThreadPool* pool_{nullptr};
+    bool owner_{false};
+};
+
+}  // namespace vsag
```
Comment (wxyucs) [inline src/safe_thread_pool.h]: add a newline at EOF

**Review by wxyucs — CHANGES_REQUESTED** [review]

File: src/safe_thread_pool.h (hunk 9 above)
Comment (inabao) [reply src/safe_thread_pool.h]: done

File: src/default_thread_pool.cpp (hunk 8 above)
Comment (inabao) [reply src/default_thread_pool.cpp]: done

File: src/default_thread_pool.cpp (hunk 4 above)
Comment (inabao) [reply src/default_thread_pool.cpp:36]: The security of this is ensured by the implementation of pool_.

File: src/safe_thread_pool.h (hunk 7 above)
Comment (wxyucs) [reply src/safe_thread_pool.h:23]: @LHT129 any suggestion?

File: src/safe_thread_pool.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    ~SafeThreadPool() override {
+        if (owner_) {
+            delete pool_;
+        }
+    }
+
+    template <class F, class... Args>
+    auto
+    enqueue(F&& f, Args&&... args)
```
Comment (LHT129) [inline src/safe_thread_pool.h]: upper case

File: src/safe_thread_pool.h [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            std::bind(std::forward<F>(f), std::forward<Args>(args)...));
+
+        std::future<return_type> res = task->get_future();
+        Enqueue([task]() { (*task)(); });
+        return res;
+    }
+
+    std::future<void>
+    Enqueue(std::function<void(void)> task) override {
+        auto func_wrapper = [task]() {
```
Comment (LHT129) [inline src/safe_thread_pool.h:56]: safe task or safe enqueue?

File: src/safe_thread_pool.h (hunk 7 above)
Comment (LHT129) [reply src/safe_thread_pool.h:23]: I agree the safe wrapper is used for threadpool behavior rather than the task

**Review by LHT129 — APPROVED** [review]

File: src/safe_thread_pool.h (hunk 11 above)
Comment (inabao) [reply src/safe_thread_pool.h:56]: safe task

File: src/safe_thread_pool.h (hunk 10 above)
Comment (inabao) [reply src/safe_thread_pool.h]: done

**Review by wxyucs — APPROVED** [review]
lgtm

**Pull request merged**